project/Boilerplate.scala

```diff
@@ -522,6 +522,12 @@ object Boilerplate {
         else
           s"def traverseN[G[_]: Applicative, Z](f: (${`A..N`}) => G[Z])(implicit traverse: Traverse[F], semigroupal: Semigroupal[F]): G[F[Z]] = Semigroupal.traverse$arity($tupleArgs)(f)"
 
+      val flatMap =
+        if (arity == 1)
+          s"def flatMap[Z](f: (${`A..N`}) => F[Z])(implicit flatMap: FlatMap[F]): F[Z] = flatMap.flatMap($tupleArgs)(f)"
+        else
+          s"def flatMapN[Z](f: (${`A..N`}) => F[Z])(implicit flatMap: FlatMap[F], semigroupal: Semigroupal[F]): F[Z] = flatMap.flatten(Semigroupal.map$arity($tupleArgs)(f))"
+
```

Comment:
Looking into the Boilerplate code I noticed, that support for `mapN` is currently implemented in this way:
- `SemigroupalArityFunctions` contains a bunch of map***X*** (***2 ≤ X ≤ 22***) functions (along with other arity functions like imap***X*** and contramap***X***.
- These implementations are used then from all other arity syntax, including `TupleSemigroupalSyntax`.

Shouldn't we follow the same pattern for `flatMapN` – i.e. add implementations for all flatMap***X*** into `SemigroupalArityFunctions` in the first place and then make use of it from everywhere else?

I am not certain about it though, just an observation.

Reply:
Also there's the `ApplyArityFunctions` trait that adds the map***X*** syntax directly to the `Apply` typeclass.

Curious, shouldn't we create a `FlatMapArityFunctions` trait (it does not exist yet) and add flatMap***X*** method directly into the `FlatMap` typeclass as well?

Reply:
Agree, jar size is a concern. Did you try by chance to measure how much the resulting jars get increased in size comparing to the base version? May be it adds not too much...

Reply:
I did not actually. I'll give it a shot today/tomorrow and see how it compares.